transportable-udfs-annotation-processor/src/main/java/com/linkedin/transport/processor/TransportProcessor.java

```diff
@@ -115,53 +113,45 @@ private void processElements(Set<? extends Element> elements) {
-            String.format("Multiple overridings of %s methods found in %s. %s",
-                TopLevelStdUDF.class.getSimpleName(),
-                elementsOverridingTopLevelStdUDFMethods.stream()
-                    .map(TypeElement::getQualifiedName)
-                    .collect(Collectors.joining(", ")),
-                Constants.MORE_THAN_ONE_OVERRIDING_ERROR),
-            udfClassElement
-        );
-      } else if (elementsOverridingTopLevelStdUDFMethods.size() == 1) {
-        String topLevelStdUdfClassName = elementsOverridingTopLevelStdUDFMethods.get(0).getQualifiedName().toString();
-        debug(String.format("TopLevelStdUDF class found: %s", topLevelStdUdfClassName));
-        _udfProperties.addUDF(topLevelStdUdfClassName, udfClassElement.getQualifiedName().toString());
-      } else {
-        // Java will guarantee that there is at least one overriding in the type hierarchy for non-abstract classes
-        // Hence control flow should never reach here
-        fatalError(String.format("No classes in type hierarchy of %s override %s methods.",
-            udfClassElement.getQualifiedName(), TopLevelStdUDF.class.getSimpleName()));
-      }
+      String topLevelStdUdfClassName =
+          elementsOverridingTopLevelStdUDFMethods.iterator().next().getQualifiedName().toString();
+      debug(String.format("TopLevelStdUDF class found: %s", topLevelStdUdfClassName));
+      _udfProperties.addUDF(topLevelStdUdfClassName, udfClassElement.getQualifiedName().toString());
     }
   }
 
-  private List<TypeElement> getParentsImplementingInterface(TypeElement typeElement,
-      TypeMirror iface) {
-    return Streams.concat(Stream.of(typeElement.getSuperclass()), typeElement.getInterfaces().stream())
+  private Stream<TypeElement> getParentsImplementingInterface(TypeElement typeElement, TypeMirror iface) {
+    return Stream.concat(Stream.of(typeElement.getSuperclass()), typeElement.getInterfaces().stream())
         .filter(parent -> _types.isAssignable(parent, iface))
-        .map(parent -> (TypeElement) _types.asElement(parent))
-        .collect(Collectors.toList());
+        .map(parent -> (TypeElement) _types.asElement(parent));
   }
 
   /**
    * Finds all elements in the type hierarchy of a {@link TypeElement} which override {@link TopLevelStdUDF} methods
    */
-  private List<TypeElement> getElementsOverridingTopLevelStdUDFMethods(TypeElement typeElement) {
-    List<TypeElement> elementsOverridingTopLevelStdUDFMethods =
-        getParentsImplementingInterface(typeElement, _topLevelStdUDFInterfaceType).stream()
-            .flatMap(parent -> getElementsOverridingTopLevelStdUDFMethods(parent).stream())
-            .distinct()
-            .collect(Collectors.toList());
+  private Stream<TypeElement> getElementsOverridingTopLevelStdUDFMethods(TypeElement typeElement) {
+    Stream<TypeElement> elementsOverridingTopLevelStdUDFMethods =
+        getParentsImplementingInterface(typeElement, _topLevelStdUDFInterfaceType)
```

Comment:
Simpler to restructure by pushing recursion down to getParentsImplementingInterface (but should change the name to reflect recursive nature) and then filtering flat result based on typeElementOverridesTopLevelStdUDFMethods.

Reply:
Done